pkg/resource/deploy/providers/registry.go

```diff
@@ -266,7 +266,18 @@ func (r *Registry) Diff(urn resource.URN, id resource.ID, olds, news resource.Pr
 
 	// Create a reference using the URN and the unknown ID and fetch the provider.
 	provider, ok := r.GetProvider(mustNewReference(urn, UnknownID))
-	contract.Assertf(ok, "'Check' must be called before 'Diff'")
+	if !ok {
+		// In this case, we are doing a diff as part of evaluating the fanout of a delete-before-replace operation.
+		// We can just use the old provider here, and we will not unload it.
+		provider, ok = r.GetProvider(mustNewReference(urn, id))
+		contract.Assertf(ok, "Provider must have been registered by NewRegistry for DBR Diff (%v::%v)", urn, id)
+
+		diff, err := provider.DiffConfig(olds, news)
+		if err != nil {
+			return plugin.DiffResult{Changes: plugin.DiffUnknown}, err
+		}
+		return diff, nil
```

Comment:
I don't follow why we return a `provider.DiffConfig` result here - aren't we trying to evaluate resource-level fanout for DBR?

Reply:
This is arguably the most critical part of the change here, I think - it gives providers the opportunity to stop a DBR cascade by saying that minor changes to their inputs won't trigger replacement. If we haven't seen the provider yet, we're assuming here that the old provider is sufficient to determine whether or not this provider resource needs to be replaced.

Reply:
That's right.

This happens when we call `Diff` on a provider resource as part of evaluating the potential changes to a resource R's dependents due to the deletion and replacement of R.